apps/whale-api/src/module.indexer/model/dftx/update.masternode.ts

```diff
@@ -0,0 +1,157 @@
+import { HexEncoder } from '../../../module.model/_hex.encoder'
+import BigNumber from 'bignumber.js'
+
+@Injectable()
+export class UpdateMasternodeIndexer extends DfTxIndexer<UpdateMasternode> {
+  OP_CODE: number = CUpdateMasternode.OP_CODE
+
+  constructor (
+    private readonly masternodeMapper: MasternodeMapper,
+    private readonly masternodeStatsMapper: MasternodeStatsMapper,
+    @Inject('NETWORK') protected readonly network: NetworkName
+  ) {
+    super()
+  }
+
+  convertAddress (address: UpdateMasternodeAddress): string | null {
+    if (address.addressPubKeyHash !== undefined) {
+      if (address.addressType === MasternodeKeyType.PKHashType) {
+        return P2PKH.to(this.network, address.addressPubKeyHash).utf8String
+      }
+      return P2WPKH.to(this.network, address.addressPubKeyHash).utf8String
+    }
+    return null
+  }
+
+  async indexTransaction (block: RawBlock, transaction: DfTxTransaction<UpdateMasternode>): Promise<void> {
+    const txn = transaction.txn
+    const data = transaction.dftx.data
+
+    let ownerAddress = null
+    let operatorAddress = null
+    for (const update of data.updates) {
+      /**
+       * // No need to handle reward address
+       * // because no rewardAddress in index
+       * 0x01 = OwnerAddress
+       * 0x02 = OperatorAddress
+       * 0x03 = SetRewardAddress
+       * 0x04 = RemRewardAddress
+       */
+      if (update.address !== undefined && update.updateType === 1) {
+        ownerAddress = txn.vout[1].scriptPubKey.addresses[0]
+      }
+      if (update.address !== undefined && update.updateType === 2) {
+        operatorAddress = this.convertAddress(update.address)
+      }
+    }
+
+    const mn = await this.masternodeMapper.get(data.nodeId)
+    if (mn !== undefined) {
+      let updateRecords: Array<{
+        height: number
+        ownerAddress: string
+        operatorAddress: string
+      }> = []
+
+      if (mn.updateRecords !== undefined) {
+        updateRecords = [...mn.updateRecords]
+      }
+
+      updateRecords = [
+        {
+          height: block.height,
+          ownerAddress: (ownerAddress !== null) ? ownerAddress : mn.ownerAddress,
+          operatorAddress: (operatorAddress !== null) ? operatorAddress : mn.operatorAddress
+        },
+        ...updateRecords
+      ]
+
+      await this.masternodeMapper.put({
+        id: data.nodeId,
+        sort: `${HexEncoder.encodeHeight(block.height)}${txn.txid}`,
+        ownerAddress: (ownerAddress !== null) ? ownerAddress : mn.ownerAddress,
+        operatorAddress: (operatorAddress !== null) ? operatorAddress : mn.operatorAddress,
+        creationHeight: block.height,
+        resignHeight: -1,
+        mintedBlocks: 0,
+        timelock: 0,
+        block: { hash: block.hash, height: block.height, medianTime: block.mediantime, time: block.time },
+        collateral: txn.vout[1].value.toFixed(8),
+        updateRecords
+      })
```

Comment:
We shouldn't be replacing all the entries since it's an update, not a new creation.

Reply:
Updating an existing record would look something like this.

https://github.com/JellyfishSDK/jellyfish/blob/bee44d910bf817e3c41d0117b0cca60b1d86dd55/apps/whale-api/src/module.indexer/model/dftx/resign.masternode.ts#L26-L30

Reply:
Updated.